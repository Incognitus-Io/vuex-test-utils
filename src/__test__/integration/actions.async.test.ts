/// <reference path="../../../types/index.d.ts"/>

import { ActionContext, ActionTree } from "vuex";

interface SyncAction extends ActionTree<any, any> {
  foobar(ctx: ActionContext<any, any>): Promise<void>;
}

describe("Integration async", () => {
  describe(".toCommit", () => {
    it("passes when the action commits a mutation", async () => {
      const actions = {
        foobar: async (ctx) => {
          await Promise.resolve();
          ctx.commit("fizzbuzz");
        },
      } as SyncAction;

      await expect.action(actions.foobar).resolves.toCommit("fizzbuzz");
    });
    it("fails when the action does not commits a mutation", async () => {
      const actions = {
        foobar: async (ctx) => {
          ctx.commit("fizzbuzz");
          await Promise.resolve();
          ctx.commit("fizz");
        },
      } as SyncAction;

      await expect(() =>
        expect.action(actions.foobar).resolves.toCommit("foobar")
      ).rejects.toThrowErrorMatchingSnapshot();
    });
  });

  describe(".toCommitAsRoot", () => {
    it("passes when the action commits mutations as root", async () => {
      const actions = {
        foobar: async (ctx) => {
          await Promise.resolve();
          ctx.commit("1", undefined, { root: true });
        },
      } as SyncAction;

      await expect.action(actions.foobar).resolves.toCommitAsRoot("1");
    });
    it("fails when the action commits mutations but not as root", async () => {
      const actions = {
        foobar: async (ctx) => {
          ctx.commit("1", undefined, { root: false });
          await Promise.resolve();
          ctx.commit("2", undefined, { root: false });
        },
      } as SyncAction;

      await expect(() =>
        expect.action(actions.foobar).resolves.toCommitAsRoot("2")
      ).rejects.toThrowErrorMatchingSnapshot();
    });
    it("fails when the action does not commits mutations", async () => {
      const actions = {
        foobar: async (ctx) => {
          ctx.commit("1", undefined, { root: true });
          await Promise.resolve();
          ctx.commit("2", undefined, { root: true });
        },
      } as SyncAction;

      await expect(() =>
        expect.action(actions.foobar).resolves.toCommitAsRoot("NotIt")
      ).rejects.toThrowErrorMatchingSnapshot();
    });
  });

  describe(".toCommitInOrder", () => {
    it("passes when the action commits mutations", async () => {
      const actions = {
        foobar: async (ctx) => {
          ctx.commit("1");
          ctx.commit("2");
          await Promise.resolve();
          ctx.commit("3");
        },
      } as SyncAction;

      await expect.action(actions.foobar).resolves.toCommitInOrder(["1", "2", "3"]);
    });
    it("passes when the acton commits some mutations when not strict", async () => {
      const actions = {
        foobar: async (ctx) => {
          ctx.commit("1");
          ctx.commit("2");
          await Promise.resolve();
          ctx.commit("3");
        },
      } as SyncAction;

      await expect.action(actions.foobar).resolves.toCommitInOrder(["1", "3"], false);
    });
    it("fails when the action does not commits mutations", async () => {
      const actions = {
        foobar: async (ctx) => {
          ctx.commit("1");
          ctx.commit("2");
          await Promise.resolve();
          ctx.commit("3");
        },
      } as SyncAction;

      await expect(() =>
        expect.action(actions.foobar).resolves.toCommitInOrder(["1", "2"])
      ).rejects.toThrowErrorMatchingSnapshot();
    });
  });

  describe(".toCommitWithPayload", () => {
    describe("strict", () => {
      it("passes when the action commits a mutation with the correct payload", async () => {
        const actions = {
          foobar: async (ctx, payload) => {
            await Promise.resolve();
            ctx.commit("fizzbuzz", payload);
          },
        } as SyncAction;

        await expect
          .action(actions.foobar, { a: "b" }).resolves
          .toCommitWithPayload("fizzbuzz", { a: "b" }, true);
      });
      it("fails when the action commits the wrong mutation mutation with the correct payload", async () => {
        const actions = {
          foobar: async (ctx, payload) => {
            await Promise.resolve();
            ctx.commit("fizzbuzz", payload);
          },
        } as SyncAction;

        await expect(async () =>
          await expect
            .action(actions.foobar, { a: "b" }).resolves
            .toCommitWithPayload("foobar", { a: "b" }, true)
        ).rejects.toThrowErrorMatchingSnapshot();
      });
      it("fails when the action commits a mutation with the wrong payload", async () => {
        const actions = {
          foobar: async (ctx, payload) => {
            await Promise.resolve();
            ctx.commit("fizzbuzz", payload);
          },
        } as SyncAction;

        await expect( async () =>
          await expect
            .action(actions.foobar, { a: "b" }).resolves
            .toCommitWithPayload("fizzbuzz", { c: "d" }, true)
        ).rejects.toThrowErrorMatchingSnapshot();
      });
    });
    describe("loose", () => {
      it("passes when the action commits a mutation with the correct payload", async () => {
        const actions = {
          foobar: async (ctx, payload) => {
            await Promise.resolve();
            ctx.commit("fizzbuzz", payload);
          },
        } as SyncAction;

        await expect
          .action(actions.foobar, { a: "b", c: "d" }).resolves
          .toCommitWithPayload("fizzbuzz", { a: "b" });
      });
      it("fails when the action commits the wrong mutation mutation with the correct payload", async () => {
        const actions = {
          foobar: async (ctx, payload) => {
            await Promise.resolve();
            ctx.commit("fizzbuzz", payload);
          },
        } as SyncAction;

        await expect(async () =>
          await expect
            .action(actions.foobar, { a: "b" }).resolves
            .toCommitWithPayload("foobar", { a: "b" }, true)
        ).rejects.toThrowErrorMatchingSnapshot();
      });
      it("fails when the action commits a mutation with the wrong payload", async () => {
        const actions = {
          foobar: async (ctx, payload) => {
            await Promise.resolve();
            ctx.commit("fizzbuzz", payload);
          },
        } as SyncAction;

        await expect(async () =>
          await expect
            .action(actions.foobar, { a: "b" }).resolves
            .toCommitWithPayload("fizzbuzz", { c: "d" })
        ).rejects.toThrowErrorMatchingSnapshot();
      });
    });
  });

  describe(".toDispatch", () => {
    it("passes when the action dispatchs a mutation", async () => {
      const actions = {
        foobar: async (ctx) => {
          await Promise.resolve();
          ctx.dispatch("fizzbuzz");
        },
      } as SyncAction;

      await expect.action(actions.foobar).resolves.toDispatch("fizzbuzz");
    });
    it("fails when the action does not dispatchs a mutation", async () => {
      const actions = {
        foobar: async (ctx) => {
          await Promise.resolve();
          ctx.dispatch("fizzbuzz");
        },
      } as SyncAction;

      await expect(async () =>
        await expect.action(actions.foobar).resolves.toDispatch("foobar")
      ).rejects.toThrowErrorMatchingSnapshot();
    });
  });

  describe(".toDispatchAsRoot", () => {
    it("passes when the action dispatchs mutations as root", async () => {
      const actions = {
        foobar: async (ctx) => {
          await Promise.resolve();
          ctx.dispatch("1", undefined, { root: true });
        },
      } as SyncAction;

      await expect.action(actions.foobar).resolves.toDispatchAsRoot("1");
    });
    it("fails when the action dispatchs mutations but not as root", async () => {
      const actions = {
        foobar: async (ctx) => {
          await Promise.resolve();
          ctx.dispatch("1", undefined, { root: false });
        },
      } as SyncAction;

      await expect(async () =>
        await expect.action(actions.foobar).resolves.toDispatchAsRoot("1")
      ).rejects.toThrowErrorMatchingSnapshot();
    });
    it("fails when the action does not dispatchs mutations", async () => {
      const actions = {
        foobar: async (ctx) => {
          await Promise.resolve();
          ctx.dispatch("1", undefined, { root: true });
        },
      } as SyncAction;

      await expect(async () =>
        await expect.action(actions.foobar).resolves.toDispatchAsRoot("NotIt")
      ).rejects.toThrowErrorMatchingSnapshot();
    });
  });

  describe(".toDispatchWithPayload", () => {
    describe("strict", () => {
      it("passes when the action dispatchs a mutation with the correct payload", async () => {
        const actions = {
          foobar: async (ctx, payload) => {
            await Promise.resolve();
            ctx.dispatch("fizzbuzz", payload);
          },
        } as SyncAction;

        await expect.action(actions.foobar, { a: "b" }).resolves.toDispatchWithPayload(
          "fizzbuzz",
          { a: "b" },
          true
        );
      });
      it("fails when the action dispatchs the wrong mutation mutation with the correct payload", async () => {
        const actions = {
          foobar: async (ctx, payload) => {
            await Promise.resolve();
            ctx.dispatch("fizzbuzz", payload);
          },
        } as SyncAction;

        await expect(async () =>
          await expect.action(actions.foobar, { a: "b" }).resolves.toDispatchWithPayload(
            "foobar",
            { a: "b" },
            true
          )
        ).rejects.toThrowErrorMatchingSnapshot();
      });
      it("fails when the action dispatchs a mutation with the wrong payload", async () => {
        const actions = {
          foobar: async (ctx, payload) => {
            await Promise.resolve();
            ctx.dispatch("fizzbuzz", payload);
          },
        } as SyncAction;

        await expect(async () =>
          await expect.action(actions.foobar, { a: "b" }).resolves.toDispatchWithPayload(
            "fizzbuzz",
            { c: "d" },
            true
          )
        ).rejects.toThrowErrorMatchingSnapshot();
      });
    });

    describe("loose", () => {
      it("passes when the action dispatchs a mutation with the correct payload", async () => {
        const actions = {
          foobar: async (ctx, payload) => {
            await Promise.resolve();
            ctx.dispatch("fizzbuzz", payload);
          },
        } as SyncAction;

        await expect.action(actions.foobar, {
          a: "b",
          c: "d",
        }).resolves.toDispatchWithPayload("fizzbuzz", { a: "b" });
      });
      it("fails when the action dispatchs the wrong mutation mutation with the correct payload", async () => {
        const actions = {
          foobar: async (ctx, payload) => {
            await Promise.resolve();
            ctx.dispatch("fizzbuzz", payload);
          },
        } as SyncAction;

        await expect(async () =>
          await expect.action(actions.foobar, { a: "b" }).resolves.toDispatchWithPayload(
            "foobar",
            { a: "b" },
            true
          )
        ).rejects.toThrowErrorMatchingSnapshot();
      });
      it("fails when the action dispatchs a mutation with the wrong payload", async () => {
        const actions = {
          foobar: async (ctx, payload) => {
            await Promise.resolve();
            ctx.dispatch("fizzbuzz", payload);
          },
        } as SyncAction;

        await expect(async () =>
          await expect.action(actions.foobar, { a: "b" }).resolves.toDispatchWithPayload("fizzbuzz", {
            c: "d",
          })
        ).rejects.toThrowErrorMatchingSnapshot();
      });
    });
  });
});
