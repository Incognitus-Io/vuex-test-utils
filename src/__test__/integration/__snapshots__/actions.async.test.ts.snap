// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration async .toCommit fails when the action does not commits a mutation 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).toCommit()[22m

Expected to have been commited:
 [32m\\"foobar\\"[39m
 but received:
 [31m[\\"fizzbuzz\\", \\"fizz\\"][39m"
`;

exports[`Integration async .toCommitAsRoot fails when the action commits mutations but not as root 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).toCommitAsRoot()[22m

Expected to have been commited as root:
 [32m{\\"options\\": {\\"root\\": true}, \\"type\\": \\"2\\"}[39m
Received:
 [31m[{\\"options\\": {\\"root\\": false}, \\"type\\": \\"1\\"}, {\\"options\\": {\\"root\\": false}, \\"type\\": \\"2\\"}][39m"
`;

exports[`Integration async .toCommitAsRoot fails when the action does not commits mutations 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).toCommitAsRoot()[22m

Expected to have been commited as root:
 [32m{\\"options\\": {\\"root\\": true}, \\"type\\": \\"NotIt\\"}[39m
Received:
 [31m[{\\"options\\": {\\"root\\": true}, \\"type\\": \\"1\\"}, {\\"options\\": {\\"root\\": true}, \\"type\\": \\"2\\"}][39m"
`;

exports[`Integration async .toCommitInOrder fails when the action does not commits mutations 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).toCommitInOrder()[22m

Expected to have been commit:
 [32m[\\"1\\", \\"2\\"][39m
Received:
 [31m[\\"1\\", \\"2\\", \\"3\\"][39m"
`;

exports[`Integration async .toCommitWithPayload loose fails when the action commits a mutation with the wrong payload 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).toCommitWithPayload()[22m

Expected to have been commited with payload:
 [32m{\\"payload\\": {\\"c\\": \\"d\\"}, \\"type\\": \\"fizzbuzz\\"}[39m
Received:
 [31m[{\\"payload\\": {\\"a\\": \\"b\\"}, \\"type\\": \\"fizzbuzz\\"}][39m"
`;

exports[`Integration async .toCommitWithPayload loose fails when the action commits the wrong mutation mutation with the correct payload 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).toCommitWithPayload()[22m

Expected to have been commited with payload:
 [32m{\\"payload\\": {\\"a\\": \\"b\\"}, \\"type\\": \\"foobar\\"}[39m
Received:
 [31m[{\\"payload\\": {\\"a\\": \\"b\\"}, \\"type\\": \\"fizzbuzz\\"}][39m"
`;

exports[`Integration async .toCommitWithPayload strict fails when the action commits a mutation with the wrong payload 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).toCommitWithPayload()[22m

Expected to have been commited with payload:
 [32m{\\"payload\\": {\\"c\\": \\"d\\"}, \\"type\\": \\"fizzbuzz\\"}[39m
Received:
 [31m[{\\"payload\\": {\\"a\\": \\"b\\"}, \\"type\\": \\"fizzbuzz\\"}][39m"
`;

exports[`Integration async .toCommitWithPayload strict fails when the action commits the wrong mutation mutation with the correct payload 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).toCommitWithPayload()[22m

Expected to have been commited with payload:
 [32m{\\"payload\\": {\\"a\\": \\"b\\"}, \\"type\\": \\"foobar\\"}[39m
Received:
 [31m[{\\"payload\\": {\\"a\\": \\"b\\"}, \\"type\\": \\"fizzbuzz\\"}][39m"
`;

exports[`Integration async .toDispatch fails when the action does not dispatchs a mutation 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).toDispatch()[22m

Expected to have been dispatched:
 [32m\\"foobar\\"[39m
 but received:
 [31m[\\"fizzbuzz\\"][39m"
`;

exports[`Integration async .toDispatchAsRoot fails when the action dispatchs mutations but not as root 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).toDispatchAsRoot()[22m

Expected to have been dispatched as root:
 [32m{\\"options\\": {\\"root\\": true}, \\"type\\": \\"1\\"}[39m
Received:
 [31m[{\\"options\\": {\\"root\\": false}, \\"type\\": \\"1\\"}][39m"
`;

exports[`Integration async .toDispatchAsRoot fails when the action does not dispatchs mutations 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).toDispatchAsRoot()[22m

Expected to have been dispatched as root:
 [32m{\\"options\\": {\\"root\\": true}, \\"type\\": \\"NotIt\\"}[39m
Received:
 [31m[{\\"options\\": {\\"root\\": true}, \\"type\\": \\"1\\"}][39m"
`;

exports[`Integration async .toDispatchWithPayload loose fails when the action dispatchs a mutation with the wrong payload 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).toDispatchWithPayload()[22m

Expected to have been dispatched with payload:
 [32m{\\"payload\\": {\\"c\\": \\"d\\"}, \\"type\\": \\"fizzbuzz\\"}[39m
Received:
 [31m[{\\"payload\\": {\\"a\\": \\"b\\"}, \\"type\\": \\"fizzbuzz\\"}][39m"
`;

exports[`Integration async .toDispatchWithPayload loose fails when the action dispatchs the wrong mutation mutation with the correct payload 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).toDispatchWithPayload()[22m

Expected to have been dispatched with payload:
 [32m{\\"payload\\": {\\"a\\": \\"b\\"}, \\"type\\": \\"foobar\\"}[39m
Received:
 [31m[{\\"payload\\": {\\"a\\": \\"b\\"}, \\"type\\": \\"fizzbuzz\\"}][39m"
`;

exports[`Integration async .toDispatchWithPayload strict fails when the action dispatchs a mutation with the wrong payload 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).toDispatchWithPayload()[22m

Expected to have been dispatched with payload:
 [32m{\\"payload\\": {\\"c\\": \\"d\\"}, \\"type\\": \\"fizzbuzz\\"}[39m
Received:
 [31m[{\\"payload\\": {\\"a\\": \\"b\\"}, \\"type\\": \\"fizzbuzz\\"}][39m"
`;

exports[`Integration async .toDispatchWithPayload strict fails when the action dispatchs the wrong mutation mutation with the correct payload 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).toDispatchWithPayload()[22m

Expected to have been dispatched with payload:
 [32m{\\"payload\\": {\\"a\\": \\"b\\"}, \\"type\\": \\"foobar\\"}[39m
Received:
 [31m[{\\"payload\\": {\\"a\\": \\"b\\"}, \\"type\\": \\"fizzbuzz\\"}][39m"
`;
